blog/2020-11-10-streaming-vitess-at-bolt.adoc

```diff
@@ -0,0 +1,181 @@
+
+==== Offset Management
+
+VStream includes a VGTID event in its response. We save the VGTID as the offset in the Kafka offset topic, so when the connector restarts, we can start from the saved VGTID. However, in rare cases when a transaction includes a huge amount of rows, VStream batches the change events into multiple responses, and only the last response has the VGTID. In such cases, we don’t have the VGTID for every change event we receive. We have a few options to solve this particular issue:
+
+* We can buffer all the change events in memory and wait for the last response that contains the VGTID to arrive. So all events will have the correct VGTID associated with them. A few disadvantages are that we’ll have higher latency before events are sent to Kafka. Also, memory usage could potentially increase quite a lot due to buffering. Buffering also adds complexity to the logic. We also have no control over the number of events VStream sends to us.
+* We can use the latest VGTID we have, which is the VGTID from the previous VStream response. If the connector fails and restarts when processing such a big transaction, it’ll restart from the VGTID of the previous VStream response, thus reprocessing some events. Therefore, it has at-least-once event delivery semantics and it expects the downstream to be idempotent. Since most transactions are not big enough, most VStream responses will have VGTID in the response, so the chance of having duplicates is low. In the end, we chose this approach for its at-least-once delivery guarantee and its design simplicity.
+
+==== Schema Management
+
+VStream’s response also includes a `FIELD` event. It’s a special event that contains the schemas of the tables of which the rows are affected. For example, if we have 2 tables, `A` and `B`. If we insert a few rows into table `A`, the `FIELD` event will only contain table `A`’s schema. The VStream is smart enough to only include the `FIELD` event whenever necessary. For example, when a VStream client reconnects, or when a table’s schema is changed.
+
+The older version of VStream includes only the column type (e.g. `Integer`, `Varchar`), no additional information such as whether the column is the primary key, whether the column has a default value, `Decimal` type’s scale and precision, `Enum` type’s allowed values, etc.
+
+The newer version (Vitess 8) of VStream starts to include more information on each column. This will help the connector to deserialize more accurately certain types and have a more precise schema in the output events to Kafka.
```

Comment:
```suggestion
The newer version (Vitess 8) of VStream starts to include more information on each column. This will help the connector to deserialize more accurately certain types and have a more precise schema in the change events sent to Kafka.
```